File: example/systems/SpriteRenderSystem.java
package net.brokenspork.systems;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;

import net.brokenspork.components.Position;
import net.brokenspork.components.Sprite;

import com.artemis.Aspect;
import com.artemis.ComponentMapper;
import com.artemis.Entity;
import com.artemis.EntitySystem;
import com.artemis.annotations.Mapper;
import com.artemis.utils.Bag;
import com.artemis.utils.ImmutableBag;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.Texture.TextureFilter;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureAtlas.AtlasRegion;
import com.badlogic.gdx.graphics.g2d.TextureRegion;

public class SpriteRenderSystem extends EntitySystem {
	@Mapper
	ComponentMapper<Position> pm;
	@Mapper
	ComponentMapper<Sprite> sm;

	private HashMap<String, AtlasRegion> regions;
	private TextureAtlas textureAtlas;
	private SpriteBatch batch;
	private OrthographicCamera camera;
	private BitmapFont font;

	private Bag<AtlasRegion> regionsByEntity;
	private List<Entity> sortedEntities;

	@SuppressWarnings("unchecked")
	public SpriteRenderSystem(OrthographicCamera camera, SpriteBatch batch) {
		super(Aspect.getAspectForAll(Position.class, Sprite.class));
		this.camera = camera;
		this.batch = batch;
	}

	@Override
	protected void initialize() {
		regions = new HashMap<String, AtlasRegion>();
		textureAtlas = new TextureAtlas(Gdx.files.internal("images-packed/pack.atlas"));
		for (AtlasRegion r : textureAtlas.getRegions()) {
			regions.put(r.name, r);
		}
		regionsByEntity = new Bag<AtlasRegion>();

		batch = new SpriteBatch();

		sortedEntities = new ArrayList<Entity>();

		Texture fontTexture = new Texture(Gdx.files.internal("fonts/normal_0.png"));
		fontTexture.setFilter(TextureFilter.Linear, TextureFilter.MipMapLinearLinear);
		TextureRegion fontRegion = new TextureRegion(fontTexture);
		font = new BitmapFont(Gdx.files.internal("fonts/normal.fnt"), fontRegion, false);
		font.setUseIntegerPositions(false);
	}

	@Override
	protected void begin() {
		batch.setProjectionMatrix(camera.combined);
		batch.begin();
	}

	@Override
	protected boolean checkProcessing() {
		return true;
	}

	@Override
	protected void processEntities(ImmutableBag<Entity> entities) {
		for (int i = 0; sortedEntities.size() > i; i++) {
			process(sortedEntities.get(i));
		}
	}

	protected void process(Entity e) {
		if (pm.has(e)) {
			Position position = pm.getSafe(e);
			Sprite sprite = sm.get(e);

			AtlasRegion spriteRegion = regionsByEntity.get(e.getId());
			batch.setColor(sprite.r, sprite.g, sprite.b, sprite.a);

			float posX = position.x - (spriteRegion.getRegionWidth() / 2 * sprite.scaleX);
			float posY = position.y - (spriteRegion.getRegionHeight() / 2 * sprite.scaleX);
			batch.draw(spriteRegion, posX, posY, 0, 0, spriteRegion.getRegionWidth(), spriteRegion.getRegionHeight(), sprite.scaleX, sprite.scaleY, sprite.rotation);
			// GdxUtils.drawCentered(batch, spriteRegion, position.x,
			// position.y);
		}
	}

	protected void end() {
		batch.end();
	}

	@Override
	protected void inserted(Entity e) {
		Sprite sprite = sm.get(e);
		regionsByEntity.set(e.getId(), regions.get(sprite.name));

		sortedEntities.add(e);

		Collections.sort(sortedEntities, new Comparator<Entity>() {
			@Override
			public int compare(Entity e1, Entity e2) {
				Sprite s1 = sm.get(e1);
				Sprite s2 = sm.get(e2);
				return s1.layer.compareTo(s2.layer);
			}
		});
	}

	@Override
	protected void removed(Entity e) {
		regionsByEntity.set(e.getId(), null);
		sortedEntities.remove(e);
	}

}

File: cocos/systems/SpriteRenderSystem.ts
module example.systems {
	
	import HashMap = artemis.utils.HashMap;
	
	import Position = example.components.Position;
	import Sprite = example.components.Sprite;
	
	import Aspect = artemis.Aspect;
	import ComponentMapper = artemis.ComponentMapper;
	import Entity = artemis.Entity;
	import EntitySystem = artemis.EntitySystem;
	import Bag = artemis.utils.Bag;
	import ImmutableBag = artemis.utils.ImmutableBag;
	import Mapper = artemis.annotations.Mapper;
	import Constants = example.core.Constants;

	export class SpriteRenderSystem extends EntitySystem {
		@Mapper(Position) pm:ComponentMapper<Position>;
		@Mapper(Sprite) sm:ComponentMapper<Sprite>;
	
		private regions:HashMap<String, cc.SpriteFrame>;
		private font:cc.LabelBMFont;
	
		private regionsByEntity:Bag<cc.SpriteFrame>;
		private sortedEntities:Array<Entity>;

    private game:CCLayer;

    constructor(game:CCLayer) {
			super(Aspect.getAspectForAll(Position, Sprite));
      this.game = game;
		}
	
		
		public initialize() {
			
			this.regions = new HashMap<String, cc.SpriteFrame>();
			cc.spriteFrameCache.addSpriteFrames("res/images-packed/pack.plist");
			
			
			var textureAtlas:any = cc.spriteFrameCache;
			for (var name in textureAtlas._spriteFrames) {
				var r = textureAtlas._spriteFrames[name];
				this.regions.put(name, r);
			}
			this.regionsByEntity = new Bag<cc.SpriteFrame>();
			
			this.sortedEntities = new Array<Entity>();
	
		}
	
		

		protected checkProcessing():boolean {
			return true;
		}
	
		
		public processEntities(entities:ImmutableBag<Entity>) {
			for (var i = 0; this.sortedEntities.length > i; i++) {
				this.processEach(this.sortedEntities[i]);
			}
		}
	
		public processEach(e:Entity) {
			if (this.pm.has(e)) {
				var position:Position = this.pm.getSafe(e);
				var sprite:Sprite = this.sm.get(e);

        sprite.sprite_.setPosition(cc.p(position.x*2, Constants.FRAME_HEIGHT - position.y));
			}
		}
	

		public inserted(e:Entity) {
			var sprite:Sprite = this.sm.get(e);

			this.regionsByEntity.set(e.getId(), this.regions.get(sprite.name));
	
			// sortedEntities.add(e);
			this.sortedEntities.push(e);
			
			this.sortedEntities.sort((e1:Entity, e2:Entity) => {
					var s1:Sprite = this.sm.get(e1);
					var s2:Sprite = this.sm.get(e2);
					return s1.layer - s2.layer;
			});
			
		}
	
		
		protected removed(e:Entity) {
      var c:Sprite = <Sprite> e.getComponentByType(Sprite);
      c.removeFrom(this.game);

			this.regionsByEntity.set(e.getId(), null);
			var index = this.sortedEntities.indexOf(e);
			if (index != -1) {
				this.sortedEntities.splice(index, 1);
			}
		}
	
	}
}

